File: android/share/src/main/kotlin/io/rynbridge/share/ShareModule.kt
package io.rynbridge.share

import android.content.Context
import io.rynbridge.core.*

class ShareModule(provider: ShareProvider) : BridgeModule {
    constructor(context: Context) : this(DefaultShareProvider(context))

    override val name = "share"
    override val version = "0.1.0"
    override val actions: Map<String, ActionHandler> = mapOf(
        "share" to { payload ->
            val text = payload["text"]?.stringValue
            val url = payload["url"]?.stringValue
            val title = payload["title"]?.stringValue
            val success = provider.share(text, url, title)
            ShareResult(success).toPayload()
        },
        "shareFile" to { payload ->
            val filePath = payload["filePath"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: filePath")
            val mimeType = payload["mimeType"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: mimeType")
            val success = provider.shareFile(filePath, mimeType)
            ShareResult(success).toPayload()
        },
        "copyToClipboard" to { payload ->
            val text = payload["text"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: text")
            provider.copyToClipboard(text)
            emptyMap()
        },
        "readClipboard" to { _ ->
            val text = provider.readClipboard()
            ClipboardText(text ?: "").toPayload()
        },
        "canShare" to { _ ->
            val canShare = provider.canShare()
            mapOf("canShare" to BridgeValue.bool(canShare))
        }
    )
}

File: ios/Sources/RynBridgeShare/ShareModule.swift
import Foundation
import RynBridge

public struct ShareModule: BridgeModule, Sendable {
    public let name = "share"
    public let version = "0.1.0"
    public let actions: [String: ActionHandler]

    public init(provider: ShareProvider) {
        actions = [
            "share": { payload in
                let text = payload["text"]?.stringValue
                let url = payload["url"]?.stringValue
                let title = payload["title"]?.stringValue
                let success = try await provider.share(text: text, url: url, title: title)
                return ShareResult(success: success).toPayload()
            },
            "shareFile": { payload in
                guard let filePath = payload["filePath"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: filePath")
                }
                guard let mimeType = payload["mimeType"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: mimeType")
                }
                let success = try await provider.shareFile(filePath: filePath, mimeType: mimeType)
                return ShareResult(success: success).toPayload()
            },
            "copyToClipboard": { payload in
                guard let text = payload["text"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: text")
                }
                try await provider.copyToClipboard(text: text)
                return [:]
            },
            "readClipboard": { _ in
                let text = try await provider.readClipboard()
                return ClipboardText(text: text ?? "").toPayload()
            },
            "canShare": { _ in
                let canShare = try await provider.canShare()
                return ["canShare": .bool(canShare)]
            },
        ]
    }
}

#if canImport(UIKit)
extension ShareModule {
    public init() {
        self.init(provider: DefaultShareProvider())
    }
}
#endif
